text/0019-stricter-validation.md

```diff
@@ -0,0 +1,143 @@
+
+### Unspecified principal/resource entity types
+
+The validator has a concept of an unspecified principal or resource entity type in an action `appliesTo` specification.
+When `principalTypes` or `resourceTypes` is omitted, we interpret the schema as declaring that the action does not apply to any particular principal/resource entity type.
+It instead applies to the "unspecified" principal or resource, which allows a users to make authorization requests using that action without providing the principal or resource,
+or when providing an entity with any of the defined entity types as the principal or resource.
+This is implemented by assigning the `AnyEntity` type to a variable when it is unspecified,
+but this depends on width subtyping between entity types which we want to eliminate from strict validation.
+
+It is not possible to extend the set of type environments to handle this case in a way that does not use `AnyEntity`.
+Constructing one type environment for each declared entity type does not soundly handle the case where an entity is _not_ provided in the query.
+If some attribute exists for every declared entity type, then an access to that attribute would be permitted, but that attribute cannot be safely accessed if an entity is not used in the query.
+We would need to further extend set of type environments to include an entity representing this case,
+but, other than `AnyEntity`, we do not have a type that can represent the type of an unspecified entity.
+
+Instead, we observe that the existence of `AnyEntity` on it's own is not a problem for strict validation,
+so we _can_ in fact use it as the type of an unspecified entity in strict mode validation.
+The `AnyEntity` type is only a problem when it allows for construction of a union type representing multiple entity types.
+We want to typecheck the policy with a distinct type environment for every entity type and for an unspecified entity.
+For each query environment, we can over-approximate by instead using `AnyEntity` as the type of the query variable.
+In each query environment, `AnyEntity` represents exactly one entity type,
+and will always represent exactly one entity type as long as the least upper bound of `AnyEntity` with every type is not defined.
+
+If we approximate every type environment with `AnyEntity`, the different type environments are now identical, so we only need to typecheck once with this approximate type environment.
+Using the `AnyEntity` types will cause the strict-mode validator to reject some expressions which might otherwise be well typed.
+For example, `principal == User::"alice"` is well typed when `principal` is `User`, and for any other entity type for `principal`, including the unspecified entity type, the expression could have type `False`, but we will not accept it.
```

Comment:
Why would we not accept an equality that always evaluate to `False`?  The text above suggests that we do allow this and then replace such equalities with the constant expression `False` in the transformation phase of the strict validator.

Reply:
We could accept it, but this is the limitation of the proposed approach. It lets us use the current typechecker for strict validation without any additional effort, but misses this case.

Reply:
I don't understand why this case is any different than other cases though.  Suppose that the type of `principal` was `Bot`.  Wouldn't we get that `principal == User::"alice"` is `False` in that case?   And if so, what's different when `principal` is `AnyEntity` instead?

Reply:
If we view `AnyEntity` as `AbitraryEntity` as you suggest, then I think this can't get type `False`. The arbitrary entity type could just as easily be `User` as any other entity type.

Reply:
Is there any disadvantage to the `ArbitraryEntity` approach?  To me, it seems like the simplest solution (both to explain and in terms of the precision it gives).  But I haven't thought through all the consequences :)

Reply:
Ah, sorry, I misread your "can't get type `False`" as "_can_ get type `False`". 

But why not?  

We have two totally different entity types, and we're comparing instances of those different entity types for equality.  Why is that not `False`?  The fact that we call one of them `ArbitraryEntity` shouldn't matter.

Reply:
To clarify, an instance of the `ArbitraryEntity` type _cannot_ be equal to an entity of any other type.  It's a distinct type that we call `ArbitraryEntity` that has no attributes.  All instances of this type will be necessarily different from all instances of other types, such as `User`.  Only two instances of `AribtraryEntity` can return  a non-False result when compared for equality.